Fix SILGen of access to isolated static properties

kavon opened the pull request:
When accessing a static property isolated to a
global-actor, such as MainActor, a `hop_to_executor`
was not being emitted. This was caught by an assertion
I had left to catch such cases, but of course in release
builds this will lead to incorrect SIL, etc.

- [x] Fix basic access of a simple stored property
- [x] Fix access of an optional reference type with chained lookup.
- [x] Prevent emission of double-hop on standard access of static global (the `getCount` case).

Resolves rdar://78292384

kavon committed:
Fix SILGen of access to isolated static properties

When accessing a static property isolated to a
global-actor, such as MainActor, a `hop_to_executor`
was not being emitted. This was caught by an assertion
I had left to catch such cases, but of course in release
builds this will lead to incorrect SIL, etc.

This issue revealed some other problems with how
the implementation was done for other kinds of
accesses starting from a static property, e.g.,
emitting a redundant hop for the same access.
This was fixed by modelling the actor-isolation
placed into a component as only being accessible
by consuming it from the component. This prevents
the emission of the hops more than once.

Thus, the regression test was updated to catch
unexpected hop_to_executor instructions.

Resolves rdar://78292384

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test

ktoso commented on lib/SILGen/SILGenLValue.cpp:
```diff
@@ -576,7 +587,8 @@ static SILValue enterAccessScope(SILGenFunction &SGF, SILLocation loc,
   std::unique_ptr<LogicalPathComponent>
-    component(new EndAccessPseudoComponent(typeData));
+    component(new EndAccessPseudoComponent(typeData,
+                                              std::move(prevExecutor)));
```
Nice, that's nicely modeled with the move and breadcrumb 👍

ktoso reviewed (approved):
Looks good to me 👍 Nicely modeled by the way.

kavon committed:
increase strictness of hop_to_executor test

Added an implicit CHECK-NOT to catch any unexpected
hop_to_executor instructions in the output of SILGen.

kavon commented:
@swift-ci please smoke test and merge

kavon commented:
@swift-ci please smoke test and merge
